docs/architecture/ADR-003-borrow-assets.md

```diff
@@ -0,0 +1,121 @@
+
+## Decision
+
+The Cosmos `x/bank` module and the existing `umee/x/leverage` deposit features are prerequisites for these new capabilities.
+
+The flow of events is as follows:
+- Borrower has already used `umee/x/leverage` module to deposit base assets and receive uTokens
+- Borrower deposits uTokens into asset facility as collateral - facility remembers collateral position
+- Borrower requests to borrow assets from capital facility - facility checks request, disburses tokens if acceptable, and remembers borrow position
+- Eventually, borrower pays repayment amount (in full or in part) using borrowed asset denomination
+- Borrower may withdraw collateral uTokens at any time as long as minimum overcollateralization requirement remains satisfied
+- On asset repayment amount = 0 for a given borrow position, facility forgets loan.
+- On uToken collateral amount = 0 for a given collateral position, facility forgets collateral.
+
+Additionally, the following events occur periodically (at the end of every block):
+- Repayment amounts for open borrow positions increases by a borrowed-token-specific interest rate
+- Open borrow positions are checked to see if they fall below liquidation threshold
+
+The `umee/x/leverage` module keeper will be responsible for remembering each open borrow or collateral position.
+If the same user account opens multiple of the same type of position (e.g. borrow) in the same token denomination, the second position simply increases the amount of the first.
+
+Additionally it has been discussed that, rather than an account's specific collateral positions being tied to specific borrow positions, the sum of all borrow and collateral positions related to an account is used to check the account's overcollateralization levels and mark its positions for liquidation.
+
+Note that the exchange rate of Assets:uAssets (e.g. `uatom:u/uatom`) will be a shifting exchange rate that grows with interest - see ADR-001.
+
+In contrast, the exchange rate of original assets:borrowed assets (e.g. `uatom:ether`) can only be determined using a price oracle.
+
+The true borrow exchange rate used in the transaction (e.g. `u/uatom:ether`) is derived from the two above.
+
+Note also that as a consequence of uToken interest, the asset value of uToken collateral increases over time, meaning a user who repays positions in full and redeems collateral uTokens will receive back more base assets than they deposited originally, reducing effective interest.
+
+## Detailed Design
+
+For the purposes of borrowing and repaying assets, as well as depositing and withdrawing uToken collateral, the `umee/x/leverage` module does not mint or burn tokens. It uses its module account to store collateral uTokens, its keeper to remember open positions, and the `x/bank` module to perform all necessary balance checks and token transfers.
```

Comment:
In the latter case, i.e. borrowing, tokens are minted and sent to the user and when withdrawing the uTokens are burnt. Not sure if this was intended in this sentence.

Reply:
(Replying as if post-discussion edits already made): No minting or burning should happen with borrow/repay transactions, as the assets being sent around are all base asset types, not uTokens. Any (loaned) assets the user receives or repays were moved from/to the module account.